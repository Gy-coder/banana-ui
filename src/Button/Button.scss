$h: 32px;
$border-color: #d9d9d9;
$color: #333;
$blue: #40a9ff;
$radius: 4px;
$red: red;
$grey: grey;
.g-button {
  box-sizing: border-box;
  padding: 0 12px;
  height: $h;
  cursor: pointer;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  white-space: nowrap;
  background: white;
  color: $color;
  border: 1px solid $border-color;
  border-radius: $radius;
  box-shadow: 0 1px 0 fade-out(black, 0.95);
  & + & {
    margin-left: 8px;
  }
  &:hover,
  &:focus {
    color: $blue;
    border-color: $blue;
  }
  &:focus {
    outline: none;
  }
  &::-moz-focus-inner {
    border: 0;
  }
  &.g-theme-link {
    border-color: transparent;
    box-shadow: none;
    color: $blue;
    &:hover,
    &:focus {
      color: lighten($blue, 10%);
    }
  }
  &.g-theme-text {
    border-color: transparent;
    box-shadow: none;
    color: inherit;
    &:hover,
    &:focus {
      background: darken(white, 5%);
    }
  }
  &.g-size-big {
    font-size: 24px;
    height: 48px;
    padding: 0 16px;
  }
  &.g-size-small {
    font-size: 12px;
    height: 20px;
    padding: 0 4px;
  }
  &.g-theme-button {
    &.g-level-main {
      background: $blue;
      color: white;
      border-color: $blue;
      &:hover,
      &:focus {
        background: darken($blue, 10%);
        border-color: darken($blue, 10%);
      }
    }
    &.g-level-danger {
      background: $red;
      border-color: $red;
      color: white;
      &:hover,
      &:focus {
        background: darken($red, 10%);
        border-color: darken($red, 10%);
      }
    }
  }
  &.g-theme-link {
    &.g-level-danger {
      color: $red;
      &:hover,
      &:focus {
        color: darken($red, 10%);
      }
    }
  }
  &.g-theme-text {
    &.g-level-main {
      color: $blue;
      &:hover,
      &:focus {
        color: darken($blue, 10%);
      }
    }
    &.g-level-danger {
      color: $red;
      &:hover,
      &:focus {
        color: darken($red, 10%);
      }
    }
  }
  &.g-theme-button {
    &[disabled] {
      cursor: not-allowed;
      color: $grey;
      &:hover {
        border-color: $grey;
      }
    }
  }
  &.g-theme-link,
  &.g-theme-text {
    &[disabled] {
      cursor: not-allowed;
      color: $grey;
    }
  }
  > .g-loadingIndicator {
    box-sizing: border-box;
    width: 14px;
    height: 14px;
    display: inline-block;
    margin-right: 4px;
    border-radius: 8px;
    border-color: $blue $blue $blue transparent;
    border-style: solid;
    border-width: 2px;
    animation: g-spin 1s infinite linear;
  }
}
@keyframes g-spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
