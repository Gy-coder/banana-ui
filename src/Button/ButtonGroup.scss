$h: 32px;
$border-color: #d9d9d9;
$color: #333;
$blue: #40a9ff;
$radius: 4px;
$red: red;
$grey: grey;
.g-button-group {
  display: inline-flex;
  vertical-align: center;
  > .g-button {
    border-radius: 0;
    margin-left: -1px !important;
    &:hover {
      z-index: 1;
    }
    &:focus {
      z-index: 1;
    }
    &:first-child {
      border-top-left-radius: $radius;
      border-bottom-left-radius: $radius;
    }
    &:last-child {
      border-top-right-radius: $radius;
      border-bottom-right-radius: $radius;
    }
  }
  > .g-button + .g-button {
    margin: 0;
  }
}
