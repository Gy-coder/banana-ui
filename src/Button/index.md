---
title: Button
nav:
  title: 基础组件
  path: /components
---

# Button

按钮一般用于点击触发一个操作

# 何时使用

标记了一个（或封装一组）操作命令，响应用户点击行为，触发相应的业务逻辑。

在 banana-ui 中我们提供了三种按钮。

- 默认按钮：用于没有主次之分的一组行动点。

- 文本按钮：用于最次级的行动点。

- 链接按钮：用于作为外链的行动点。

提供了大中小三个不同的尺寸的按钮

以及四种状态属性与上面配合使用。

- 危险：删除/移动/修改权限等危险操作，一般需要二次确认。

- 主按钮: 用于主行动点，一个操作区域只能有一个主按钮。

- 禁用：行动点不可用的时候，一般需要文案解释。

- 加载中：用于异步操作等待反馈的时候，也可以避免多次提交。

# 代码演示

<code src="./demos/basic.tsx" />
<code src="./demos/size.tsx" />
<code src="./demos/level.tsx" />
<code src="./demos/disabled.tsx">
<code src="./demos/loading.tsx" />

<API src="./Button.tsx"/>
