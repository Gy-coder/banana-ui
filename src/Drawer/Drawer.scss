.g-drawer {
  position: fixed;
  z-index: 1000;
  width: 0%;
  height: 100%;
  transition: width 0s ease 0.3s, height 0s ease 0.3s;

  &-mask {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 0;
    background-color: rgba(0, 0, 0, 0.45);
    opacity: 0;
    transition: opacity 0.3s linear, height 0s ease 0.3s;
  }

  &-content {
    position: absolute;
    background: white;
    width: 100%;
    height: 100%;
    transition: all 0.3s cubic-bezier(0.23, 1, 0.32, 1);
    display: flex;
    > .g-drawer-header {
      position: relative;
      display: flex;
      align-items: center;
      justify-content: space-between;
      padding: 16px 24px;
      color: rgba(0, 0, 0, 0.85);
      background: #fff;
      border-bottom: 1px solid #f0f0f0;
      border-radius: 2px 2px 0 0;
    }
    > .g-drawer-main {
      flex-grow: 1;
      padding: 24px;
      overflow: auto;
      font-size: 14px;
      line-height: 1.5715;
      word-wrap: break-word;
      overflow: auto;
    }
    > .g-drawer-footer {
      flex-shrink: 0;
      padding: 10px 16px;
      border-top: 1px solid #f0f0f0;
    }
  }

  &.g-drawer-close {
    &.g-drawer-right {
      > .g-drawer-content {
        transform: translateX(100%);
      }
    }
  }

  &.g-drawer-open {
    width: 100%;
    > .g-drawer-mask {
      height: 100%;
      opacity: 1;
    }
  }
}

.g-drawer-right,
.g-drawer-left {
  top: 0;
  width: 0%;
  height: 100%;
  .g-drawer-content {
    flex-direction: column;
  }
}

.g-drawer-right {
  right: 0;
  > .g-drawer-content {
    right: 0;
  }
}
