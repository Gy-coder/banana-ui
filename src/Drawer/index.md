---
title: Drawer
nav:
  title: Drawer
  path: /components
---

# Drawer

屏幕边缘滑出的浮层面板。

# 何时使用

抽屉从父窗体边缘滑入，覆盖住部分父窗体内容。用户在抽屉内操作时不必离开当前任务，操作完成后，可以平滑地回到原任务。

- 当需要一个附加的面板来控制父窗体内容，这个面板在需要时呼出。比如，控制界面展示样式，往界面中添加内容。

- 当需要在当前任务流中插入临时任务，创建或预览附加内容。比如展示协议条款，创建子对象。

# 代码范例

## 基础使用

<code src="./demos/basic.tsx" />

<code src="./demos/dontShowMask.tsx" />

## API

| 参数         | 说明                                       | 类型                                   | 是否必须 | 默认值        |
| ------------ | ------------------------------------------ | -------------------------------------- | -------- | ------------- |
| visible      | 是否显示抽屉                               | boolean                                | true     |               |
| onClose      | 关闭函数                                   | ()=> void                              | true     |               |
| closeable    | 抽屉是否可以关闭                           | boolean                                | false    | true          |
| container    | 抽屉挂载的节点，false 为挂载在当前 dom     | HTMLElement ｜ false;                  | false    | document.body |
| placement    | 抽屉的方向                                 | "top" ｜ "left" ｜ "bottom" ｜ "right" | false    | "right"       |
| height       | 高度, 在 placement 为 top 或 bottom 时使用 | number                                 | false    | 384           |
| width        | 宽度 在 placement 为 left 或 right 时使用  | number                                 | false    | 384           |
| zIndex       | 抽屉的 z-index 值                          | nubmer                                 | false    | 1000          |
| showMask     | 是否显示遮罩层                             | boolean                                | false    | true          |
| maskClosable | 点击遮罩层是否可以关闭抽屉                 | boolean                                | false    | true          |
| title        | 标题                                       | ReactNode                              | false    |               |
| footer       | 抽屉的页脚                                 | ReactNode                              | false    |               |
| headerStyle  | 用于设置 Drawer 头部的样式                 | CSSProperties                          | false    |               |
| mainStyle    | 用于设置 Drawer 主区域的样式               | CSSProperties                          | false    |               |
| footerStyle  | 抽屉页脚部件的样式                         | CSSProperties                          | false    |               |
