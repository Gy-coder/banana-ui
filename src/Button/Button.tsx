import React from 'react';
import classnames from 'classnames';
import './Button.scss';

interface ButtonProps
  extends Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'disabled'> {
  /**
   * 可以这样写属性描述
   * @description     Button的类型
   * @default         button
   */
  theme?: 'button' | 'link' | 'text';
  /**
   * 可以这样写属性描述
   * @description      Button的大小
   * @default          normal
   */
  size?: 'small' | 'normal' | 'big';
  /**
   * 可以这样写属性描述
   * @description      Button的级别
   * @default          normal
   */
  level?: 'main' | 'danger' | 'normal';
  /**
   * 可以这样写属性描述
   * @description      是否禁用
   * @default          false
   */
  disabled?: boolean;
  /**
   * 可以这样写属性描述
   * @description      是否显示加载状态
   * @default          false
   */
  loading?: Boolean;
}

const Button: React.FC<ButtonProps> = (props) => {
  const {
    children,
    className,
    theme,
    size,
    level,
    disabled,
    loading,
    ...rest
  } = props;
  const classes = classnames('g-button', className, {
    [`g-theme-${theme}`]: theme,
    [`g-size-${size}`]: size,
    [`g-level-${level}`]: level,
  });
  return (
    <button className={classes} disabled={disabled} {...rest}>
      {loading && <span className="g-loadingIndicator" />}
      {children}
    </button>
  );
};

Button.defaultProps = {
  theme: 'button',
  size: 'normal',
  level: 'normal',
  disabled: false,
  loading: false,
};

export default Button;
