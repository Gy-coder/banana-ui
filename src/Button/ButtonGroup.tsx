import React from 'react';
import Button, { ButtonProps } from './Button';
import './ButtonGroup.scss';

const ButtonGroup: React.FC = (props) => {
  const { children } = props;
  React.Children.map(children, (child) => {
    const childElement = child as React.FunctionComponentElement<ButtonProps>;
    if (childElement.type !== Button) {
      throw new Error('CheckboxGroup的子元素必须是Checkbox');
    }
  });
  return <div className="g-button-group">{children}</div>;
};

export default ButtonGroup;
