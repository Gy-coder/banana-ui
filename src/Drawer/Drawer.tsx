import React, { CSSProperties, ReactNode, useEffect, useRef } from 'react';
import ReactDOM from 'react-dom';
import classnames from 'classnames';
import { CloseCircleOutlined } from '@ant-design/icons';
import './Drawer.scss';

export interface DrawerProps {
  visible: boolean;
  onClose: () => void;
  closeable?: boolean;
  container?: HTMLElement | false;
  placement?: 'top' | 'left' | 'bottom' | 'right';
  width?: number;
  zIndex?: number;
  showMask?: boolean;
  maskClosable?: boolean;
  title?: ReactNode;
  footer?: ReactNode;
  headerStyle?: CSSProperties;
  mainStyle?: CSSProperties;
  footerStyle?: CSSProperties;
}

const Drawer: React.FC<DrawerProps> = (props) => {
  const cn = 'g-drawer';
  const {
    visible,
    onClose,
    closeable,
    placement = 'top',
    container,
    width,
    zIndex,
    showMask,
    maskClosable,
    title,
    footer,
    headerStyle,
    mainStyle,
    footerStyle,
    children,
  } = props;
  const componentRef = useRef<HTMLDivElement>(null);
  const handleClose = () => {
    onClose();
  };

  useEffect(() => {
    if (visible && showMask) document.body.style.overflow = 'hidden';
    else document.body.style.overflow = 'auto';
  }, [visible]);
  const html = () => {
    return (
      <div
        className={classnames(cn, cn + `-${placement}`, {
          'g-drawer-open': visible,
          'g-drawer-close': !visible,
        })}
        ref={componentRef}
        style={{
          zIndex,
        }}
      >
        {!!showMask && (
          <div
            className={cn + '-mask'}
            onClick={maskClosable ? handleClose : undefined}
          />
        )}
        <div
          className={classnames(cn + '-content')}
          style={{
            zIndex,
            width,
          }}
        >
          <div className={cn + '-header'} style={headerStyle}>
            {title}
            {closeable && (
              <CloseCircleOutlined
                className="close-button"
                onClick={handleClose}
              />
            )}
          </div>
          <div className={cn + '-main'} style={mainStyle}>
            {children}
          </div>
          {footer && (
            <div className={cn + '-footer'} style-={footerStyle}>
              {footer}
            </div>
          )}
        </div>
      </div>
    );
  };

  return container === false
    ? html()
    : ReactDOM.createPortal(html(), document.body);
};

Drawer.defaultProps = {
  closeable: true,
  placement: 'right',
  width: 384,
  zIndex: 1000,
  showMask: true,
  maskClosable: true,
  container: document.body,
};

export default Drawer;
