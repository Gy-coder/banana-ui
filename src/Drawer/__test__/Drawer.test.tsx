import {
  act,
  render,
  cleanup,
  screen,
  fireEvent,
  RenderResult,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import React, { useState } from 'react';
import Drawer, { DrawerProps } from '../Drawer';
import Button from '../../Button/Button';

let visible: boolean, setVisible: Function;
const Demo = () => {
  [visible, setVisible] = useState(false);
  const handleClose = () => {
    setVisible(false);
  };
  return (
    <>
      <Button onClick={() => setVisible(true)}>show</Button>
      <Drawer
        placement="right"
        visible={visible}
        onClose={handleClose}
        title="这是一个Drawer"
        footer="footer"
      >
        <p>title</p>
      </Drawer>
    </>
  );
};

let wrapper: RenderResult, btn: HTMLButtonElement;

describe('Testing Drawer component', () => {
  it('Drawer should be render correct', async () => {
    wrapper = render(<Demo />);
    btn = screen.getByText('show');
    expect(btn).toBeInTheDocument();
    fireEvent.click(btn);
    expect(visible).toBeTruthy();
    expect(screen.queryByText('title')).toBeInTheDocument();
    expect(wrapper.asFragment()).toMatchSnapshot();
    expect(
      wrapper.container.querySelector('.g-drawer.g-drawer-right.g-drawer-open'),
    ).toBeInTheDocument();
    expect(wrapper.asFragment()).toMatchSnapshot();
  });
  it('should close', async () => {
    wrapper = render(<Demo />);
    const closeBtn = document.getElementsByClassName('close-button')[0];
    fireEvent.click(closeBtn);
    await waitForElementToBeRemoved(() => screen.queryAllByText('title'));
  });
});
