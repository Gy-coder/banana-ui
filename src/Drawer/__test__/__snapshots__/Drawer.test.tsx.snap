// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Drawer component Drawer should be render correct 1`] = `
<DocumentFragment>
  <button
    class="g-button g-theme-button g-size-normal g-level-normal"
  >
    show
  </button>
</DocumentFragment>
`;

exports[`Testing Drawer component Drawer should be render correct 2`] = `
<DocumentFragment>
  <button
    class="g-button g-theme-button g-size-normal g-level-normal"
  >
    show
  </button>
</DocumentFragment>
`;
